x-pack/plugins/security_solution/public/common/links/links.tsx

```diff
@@ -0,0 +1,199 @@
+import {
+  Feature,
+  LinkInfo,
+  LinkItem,
+  LinkProps,
+  NavLinkItem,
+  NormalizedLink,
+  NormalizedLinks,
+} from './types';
+
+const createDeepLink = (link: LinkItem, linkProps?: LinkProps): AppDeepLink => ({
+  id: link.id,
+  path: link.path,
+  title: link.title,
+  ...(link.links && link.links.length
+    ? {
+        deepLinks: reduceLinks<AppDeepLink>({
+          links: link.links,
+          linkProps,
+          formatFunction: createDeepLink,
+        }),
+      }
+    : {}),
+  ...(link.icon != null ? { euiIconType: link.icon } : {}),
+  ...(link.image != null ? { icon: link.image } : {}),
+  ...(link.globalSearchKeywords != null ? { keywords: link.globalSearchKeywords } : {}),
+  ...(link.globalNavEnabled != null
+    ? { navLinkStatus: link.globalNavEnabled ? AppNavLinkStatus.visible : AppNavLinkStatus.hidden }
+    : {}),
+  ...(link.globalNavOrder != null ? { order: link.globalNavOrder } : {}),
+  ...(link.globalSearchEnabled != null ? { searchable: link.globalSearchEnabled } : {}),
+});
+
+const createNavLinkItem = (link: LinkItem, linkProps?: LinkProps): NavLinkItem => ({
+  id: link.id,
+  path: link.path,
+  title: link.title,
+  ...(link.description != null ? { description: link.description } : {}),
+  ...(link.icon != null ? { icon: link.icon } : {}),
+  ...(link.image != null ? { image: link.image } : {}),
+  ...(link.links && link.links.length
+    ? {
+        links: reduceLinks<NavLinkItem>({
+          links: link.links,
+          linkProps,
+          formatFunction: createNavLinkItem,
+        }),
+      }
+    : {}),
+  ...(link.skipUrlState != null ? { skipUrlState: link.skipUrlState } : {}),
+});
+
+const hasFeaturesCapability = (
+  features: Feature[] | undefined,
+  capabilities: Capabilities
+): boolean => {
+  if (!features) {
+    return true;
+  }
+  return features.some((featureKey) => get(capabilities, featureKey, false));
+};
+
+export function reduceLinks<T>({
+  links,
+  linkProps,
+  formatFunction,
+}: {
+  links: LinkItem[];
+  linkProps?: LinkProps;
+  formatFunction: (link: LinkItem, linkProps?: LinkProps) => T;
+}): T[] {
+  return links.reduce((deepLinks: T[], link: LinkItem) => {
+    if (
+      linkProps != null &&
+      // exclude link when license is basic and link is premium
+      ((linkProps.isBasic && link.isPremium) ||
+        // exclude link when enableExperimental[hideWhenExperimentalKey] is enabled and link has hideWhenExperimentalKey
+        (link.hideWhenExperimentalKey != null &&
+          linkProps.enableExperimental[link.hideWhenExperimentalKey]) ||
+        // exclude link when enableExperimental[experimentalKey] is disabled and link has experimentalKey
+        (link.experimentalKey != null && !linkProps.enableExperimental[link.experimentalKey]) ||
+        // exclude link when link is not part of enabled feature capabilities
+        (linkProps.capabilities != null &&
+          !hasFeaturesCapability(link.features, linkProps.capabilities)))
```

Comment:
I think we can assume that if the linkProps is passed, it will contain everything right?. Also, would it be possible to extract all these conditionals into a: 
`const isLinkAllowed = (link, { licenseType, enableExperimental, capabilities }) => boolean` or something?

Reply:
capabilities can be undefined but i will adjust the rest